library/cwm/test/cwm_test.rb

```diff
@@ -0,0 +1,229 @@
+#! /usr/bin/env rspec --format doc
+
+require_relative "test_helper"
+
+require "yast"
+Yast.import "CWM"
+
+# The handler methods must exist because fun_ref needs that
+# but they are stubbed out.
+module MyHandlers
+  def w1_init(_key)
+    raise
+  end
+
+  def w1_handle(_key, _event)
+    raise
+  end
+
+  def w1_validate(_key, _event)
+    raise
+  end
+
+  def w2_store(_key, _event)
+    raise
+  end
+
+  def w2_handle(_key, _event)
+    raise
+  end
+
+  def w2_validate(_key, _event)
+    raise
+  end
+
+  def generic_init(_key)
+    raise
+  end
+
+  def generic_save(_key, _event)
+    raise
+  end
+end
+
+describe Yast::CWMClass do
+  subject { Yast::CWM }
+  include Yast
+  include MyHandlers
+
+  let(:test_stringterm) { VBox(HBox("w1")) }
+  let(:widget_names) { ["w1", "w2"] }
+  let(:test_widgets) do
+    {
+      "w1" => {
+        "widget"            => :checkbox,
+        "opt"               => [:notify, :immediate],
+        "label"             => "Check&Box",
+        "init"              => fun_ref(method(:w1_init), "void (string)"),
+        "handle"            => fun_ref(
+          method(:w1_handle),
+          "symbol (string, map)"
+        ),
+        "validate_type"     => :function,
+        "validate_function" => fun_ref(
+          method(:w1_validate),
+          "boolean (string, map)"
+        )
+      },
+      "w2" => {
+        "widget"            => :textentry,
+        "label"             => "Text&Entry",
+        "store"             => fun_ref(
+          method(:w2_store),
+          "void (string, map)"
+        ),
+        "handle"            => fun_ref(
+          method(:w2_handle),
+          "symbol (string, map)"
+        ),
+        "validate_type"     => :function,
+        "validate_function" => fun_ref(
+          method(:w2_validate),
+          "boolean (string, map)"
+        )
+      }
+    }
+  end
+
+  let(:fallback_funcs) do
+    {
+      "init"  => fun_ref(method(:generic_init), "void (string)"),
+      "store" => fun_ref(method(:generic_save), "void (string, map)")
+    }
+  end
+
+  let(:created_widgets) { subject.CreateWidgets(widget_names, test_widgets) }
+  let(:run_widgets) { subject.mergeFunctions(created_widgets, fallback_funcs) }
+
+  # many public uses
+  describe "#CreateWidgets" do
+    let(:cw) { created_widgets }
+
+    it "produces an Array" do
+      expect(cw).to be_an(Array)
+    end
+
+    it "creates Terms at the 'widget' keys" do
+      expect(cw[0]["widget"]).to eq(CheckBox(Id("w1"), Opt(:notify, :immediate), "Check&Box"))
+      expect(cw[1]["widget"]).to eq(InputField(Id("w2"), Opt(:hstretch), "Text&Entry"))
+    end
+
+    # used by other CWM modules+classes
+    it "creates the '_cwm_key' keys" do
+      expect(cw[0]["_cwm_key"]).to eq("w1")
+      expect(cw[1]["_cwm_key"]).to eq("w2")
+    end
+  end
+
+  # used by CWMTab
```

Comment:
please check, maybe CWMTab is already dead? Originally it was created for bootloader, but there we use object version.

Reply:
Not dead. 16 matches for https://github.com/search?q=org%3Ayast+CWMTab.CreateWidget&type=Code